### PR TITLE
Developed update status database function and support functions.

### PR DESCRIPTION
**Pull request by yashodgayashan**
### Background Information
* Every and each application has assignments section which hold the information about the assign ministries and its status. 
* When new status is updated it will create a new version of the status and push it to the status array in appropriate assignment in assignments array.

### CODE REVIEW

File: core/src/env_platform/utility.bal [hunk 1]
```diff
@@ -203,3 +203,102 @@ function constructAssignmentArray(AssignedMinistry assignedMinistry, json[] assi
...
+            if (progress == "Completed") {
+                return true;
+            }
+            break;
+        }
+    }
+    return false;
+}
+
+# The `updateAssignment` function will push the incomming status to the status array of the assignment.
```
Comment (sanjulamadurapperuma) [inline core/src/env_platform/utility.bal:254]:
```suggestion
# The `updateAssignment` function will push the incoming status to the status array of the assignment.
```

File: core/src/env_platform/utility.bal [hunk 2]
```diff
@@ -203,3 +203,102 @@ function constructAssignmentArray(AssignedMinistry assignedMinistry, json[] assi
...
+# + status - Incomming status.
+# + return - This function will return updated assignment or an error if occured.
+function updateAssignment(json assignment, Status status) returns json|error {
+    map<json> assignmentInfo = check trap <map<json>>assignment;
+    json[] statusArray = check trap <json[]>assignmentInfo.status;
+    statusArray.push(check constructStatus(status));
+    return assignmentInfo;
+}
+
+# The `custructStatus` function will construct the given status to a appropriate format.
```
Comment (sanjulamadurapperuma) [inline core/src/env_platform/utility.bal:266]:
```suggestion
# The `constructStatus` function will construct the given status to a appropriate format.
```

File: core/src/env_platform/utility.bal [hunk 3]
```diff
@@ -203,3 +203,102 @@ function constructAssignmentArray(AssignedMinistry assignedMinistry, json[] assi
...
+            timestamp: check json.constructFrom(status.timestamp),
+            reason: status?.reason
+        };
+    }
+}
+
+# The `updateAssignments` function will updated the assignment array with updated assignment.
+# 
+# + assignments - Array of assignments of the application.
+# + updatedAssignment - Assignment which updated with incomming status.
```
Comment (sanjulamadurapperuma) [inline core/src/env_platform/utility.bal:291]:
```suggestion
# + updatedAssignment - Assignment which is updated with incoming status.
```

File: core/src/env_platform/utility.bal [hunk 4]
```diff
@@ -203,3 +203,102 @@ function constructAssignmentArray(AssignedMinistry assignedMinistry, json[] assi
...
+# + ministryId - Id of the ministry which assignment should be altered.
+# + return - This function will return null or not exsits error.
```
Comment (sanjulamadurapperuma) [inline core/src/env_platform/utility.bal:293]:
```suggestion
# + return - This function will return null or not exist error.
```

**Comment by yashodgayashan** [comment]
@sanjulamadurapperuma resolved requested changes.

File: core/src/env_platform/db_handler.bal [hunk 5]
```diff
@@ -350,3 +350,38 @@ function isMinistry(string ministryId) returns boolean|error {
...
+    if (applications.length() == 0) {
+        return error("Not found", message = "Application with application Id: " + applicationId + " is not found.");
+    } else {
+        map<json> application = check trap <map<json>>applications[0];
+        json[]|error assignments = trap <json[]>application.assignments;
+        if (assignments is json[]) {
+            [boolean, json?, boolean, boolean] [isMinistryAssigned, assignment, hasPrerequisite, isPrerequisiteCompeted] = check getAssignedMinistryInfo(assignments, status.ministry.id);
+            if (isMinistryAssigned) {
+                if (!hasPrerequisite || (hasPrerequisite && isPrerequisiteCompeted)){
+                    // Update the assignment with new status
```
Comment (deshankoswatte) [inline core/src/env_platform/db_handler.bal:371]:
```suggestion
                    // Update the assignment with new status.
```

File: core/src/env_platform/db_handler.bal [hunk 6]
```diff
@@ -350,3 +350,38 @@ function isMinistry(string ministryId) returns boolean|error {
...
+                    json updatedAssignment = check updateAssignment(assignment, status);
+                    // Update the assignments array
```
Comment (deshankoswatte) [inline core/src/env_platform/db_handler.bal:373]:
```suggestion
                    // Update the assignments array.
```

File: core/src/env_platform/db_handler.bal [hunk 7]
```diff
@@ -350,3 +350,38 @@ function isMinistry(string ministryId) returns boolean|error {
...
+                    check updateAssignments(assignments,updatedAssignment, status.ministry.id);
+                    int updated = check applicationCollection->update({assignments: assignments}, {"applicationId": applicationId});
+                    return updated==1? true: false;
```
Comment (deshankoswatte) [inline core/src/env_platform/db_handler.bal:376]:
```suggestion
                    return updated == 1 ? true : false; 
```

File: core/src/env_platform/db_handler.bal [hunk 8]
```diff
@@ -350,3 +350,38 @@ function isMinistry(string ministryId) returns boolean|error {
...
+            if (isMinistryAssigned) {
+                if (!hasPrerequisite || (hasPrerequisite && isPrerequisiteCompeted)){
+                    // Update the assignment with new status
+                    json updatedAssignment = check updateAssignment(assignment, status);
+                    // Update the assignments array
+                    check updateAssignments(assignments,updatedAssignment, status.ministry.id);
+                    int updated = check applicationCollection->update({assignments: assignments}, {"applicationId": applicationId});
+                    return updated==1? true: false; 
+                }else{
+                    return error("Prerequisite Ministry approval pending", message="Prerequisite ministry has not completed processing the application");
```
Comment (deshankoswatte) [inline core/src/env_platform/db_handler.bal:378]:
```suggestion
                    return error("Prerequisite Ministry approval pending", message="Prerequisite ministry has not completed processing the application.");
```

File: core/src/env_platform/utility.bal [hunk 9]
```diff
@@ -203,3 +203,102 @@ function constructAssignmentArray(AssignedMinistry assignedMinistry, json[] assi
...
+        }
+    }
+    return false;
+}
+
+# The `updateAssignment` function will push the incoming status to the status array of the assignment.
+# 
+# + assignment - Single ministry assignment.
+# + status - Incomming status.
+# + return - This function will return updated assignment or an error if occured.
```
Comment (deshankoswatte) [inline core/src/env_platform/utility.bal:258]:
```suggestion
# + return - This function will return updated assignment or an error if occurred.
```

File: core/src/env_platform/utility.bal [hunk 10]
```diff
@@ -203,3 +203,102 @@ function constructAssignmentArray(AssignedMinistry assignedMinistry, json[] assi
...
+    map<json> assignmentInfo = check trap <map<json>>assignment;
+    json[] statusArray = check trap <json[]>assignmentInfo.status;
+    statusArray.push(check constructStatus(status));
+    return assignmentInfo;
+}
+
+# The `constructStatus` function will construct the given status to a appropriate format.
+# 
+# + status - Status which should be formatted.
+# + return - This function will return either formatted json or approprate error.
```
Comment (deshankoswatte) [inline core/src/env_platform/utility.bal:269]:
```suggestion
# + return - This function will return either formatted json or an appropriate error.
```

File: core/src/env_platform/utility.bal [hunk 11]
```diff
@@ -203,3 +203,102 @@ function constructAssignmentArray(AssignedMinistry assignedMinistry, json[] assi
...
+# + status - Incomming status.
+# + return - This function will return updated assignment or an error if occured.
+function updateAssignment(json assignment, Status status) returns json|error {
+    map<json> assignmentInfo = check trap <map<json>>assignment;
+    json[] statusArray = check trap <json[]>assignmentInfo.status;
+    statusArray.push(check constructStatus(status));
+    return assignmentInfo;
+}
+
+# The `constructStatus` function will construct the given status to a appropriate format.
```
Comment (deshankoswatte) [inline core/src/env_platform/utility.bal:266]:
```suggestion
# The `constructStatus` function will construct the given status to an appropriate format.
```

File: core/src/env_platform/utility.bal [hunk 12]
```diff
@@ -203,3 +203,102 @@ function constructAssignmentArray(AssignedMinistry assignedMinistry, json[] assi
...
+                hasPrerequisite = true;
+                isPrerequisiteCompeted = check isMinistryCompleted(assignments, prerequisiteId);
+            }
+            break;
+        }
+    }
+    return [isMinistryAssigned, assignmentInfo, hasPrerequisite, isPrerequisiteCompeted];
+}
+
+# The `isMinistryCompleted` function will return whether particular ministry is complete processing application.
```
Comment (deshankoswatte) [inline core/src/env_platform/utility.bal:235]:
```suggestion
# The `isMinistryCompleted` function will return whether a particular ministry has completed processing the application.
```

File: core/src/env_platform/utility.bal [hunk 13]
```diff
@@ -203,3 +203,102 @@ function constructAssignmentArray(AssignedMinistry assignedMinistry, json[] assi
...
+}
+
+# The `updateAssignments` function will updated the assignment array with updated assignment.
+# 
+# + assignments - Array of assignments of the application.
+# + updatedAssignment - Assignment which is updated with incoming status.
+# + ministryId - Id of the ministry which assignment should be altered.
+# + return - This function will return null or not exist error.
+function updateAssignments(json[] assignments, json updatedAssignment, string ministryId) returns error?{
+
```
Comment (deshankoswatte) [inline core/src/env_platform/utility.bal:295]:
```suggestion
```

File: core/src/env_platform/utility.bal [hunk 14]
```diff
@@ -203,3 +203,102 @@ function constructAssignmentArray(AssignedMinistry assignedMinistry, json[] assi
...
+    statusArray.push(check constructStatus(status));
+    return assignmentInfo;
+}
+
+# The `constructStatus` function will construct the given status to a appropriate format.
+# 
+# + status - Status which should be formatted.
+# + return - This function will return either formatted json or approprate error.
+function constructStatus(Status status) returns json|error {
+
```
Comment (deshankoswatte) [inline core/src/env_platform/utility.bal:271]:
```suggestion
```

**Comment by yashodgayashan** [comment]
@deshankoswatte resolved requested changes

File: core/src/env_platform/db_handler.bal [hunk 15]
```diff
@@ -350,3 +350,38 @@ function isMinistry(string ministryId) returns boolean|error {
...
+            [boolean, json?, boolean, boolean] [isMinistryAssigned, assignment, hasPrerequisite, isPrerequisiteCompeted] = check getAssignedMinistryInfo(assignments, status.ministry.id);
+            if (isMinistryAssigned) {
+                if (!hasPrerequisite || (hasPrerequisite && isPrerequisiteCompeted)){
+                    // Update the assignment with new status.
+                    json updatedAssignment = check updateAssignment(assignment, status);
+                    // Update the assignments array.
+                    check updateAssignments(assignments,updatedAssignment, status.ministry.id);
+                    int updated = check applicationCollection->update({assignments: assignments}, {"applicationId": applicationId});
+                    return updated == 1 ? true : false; 
+                }else{
```
Comment (sanjulamadurapperuma) [inline core/src/env_platform/db_handler.bal:377]:
```suggestion
                } else{
```

File: core/src/env_platform/utility.bal [hunk 16]
```diff
@@ -203,3 +203,100 @@ function constructAssignmentArray(AssignedMinistry assignedMinistry, json[] assi
...
+# The `updateAssignments` function will updated the assignment array with updated assignment.
+# 
+# + assignments - Array of assignments of the application.
+# + updatedAssignment - Assignment which is updated with incoming status.
+# + ministryId - Id of the ministry which assignment should be altered.
+# + return - This function will return null or not exist error.
+function updateAssignments(json[] assignments, json updatedAssignment, string ministryId) returns error?{
+    int id = 0;
+    foreach json assignment in assignments{
```
Comment (sanjulamadurapperuma) [inline core/src/env_platform/utility.bal:295]:
```suggestion
    foreach json assignment in assignments {
```

File: core/src/env_platform/utility.bal [hunk 17]
```diff
@@ -203,3 +203,100 @@ function constructAssignmentArray(AssignedMinistry assignedMinistry, json[] assi
...
+        if(check trap assignment.id==ministryId){
```
Comment (sanjulamadurapperuma) [inline core/src/env_platform/utility.bal:296]:
```suggestion
        if (check trap assignment.id==ministryId) {
```

**Pull request merged**